common/common.go

```diff
@@ -645,3 +645,59 @@ func (c *Counter) IncrementAndGet() int64 {
 	}
 	return newID
 }
+
+// ElementProcessor defines the function signature for processing an individual element with its index.
+type ElementProcessor[E any] func(index int, element E) error
+
+// ProcessElementsByBatch takes a slice of elements and processes them in batches of `batchSize` concurrently.
+// For example, if batchSize = 10 and list has 100 elements, 10 goroutines will process 10 elements concurrently
+// in each batch. Each batch completes before the next batch begins.
+// `process` is a function called for each individual element with its index and value.
+func ProcessElementsByBatch[S ~[]E, E any](batchSize int, list S, process ElementProcessor[E]) (errs error) {
+	var wg sync.WaitGroup
+	errC := make(chan error, len(list))
+	for i, s := range Batch(list, batchSize) {
+		wg.Add(len(s))
+		for j, e := range s {
+			go func(index int, element E) {
+				defer wg.Done()
+				if err := process(index, element); err != nil {
+					errC <- err
+				}
+			}((i*batchSize)+j, e)
+		}
+		wg.Wait()
+	}
+	close(errC)
+	for err := range errC {
+		errs = AppendError(errs, err)
+	}
+	return errs
+}
+
+// BatchProcessor defines the function signature for processing a batch of elements.
+type BatchProcessor[S ~[]E, E any] func(batch S) error
+
+// ProcessBatches takes a slice of elements and processes them in batches of `batchSize` concurrently.
+// For example, if batchSize = 10 and list has 100 elements, 10 goroutines will be created to process
+// 10 batches. Each batch is processed sequentially by the `process` function, and batches are processed
+// in parallel.
```

Comment:
I'm afraid I had to re-read this comment a few times.
I don't believe we should say:
```Each batch is processed sequentially by the `process` function```
Because I don't think you can guarantee the goros will execute in the order scheduled.

I'd suggest vastly simplified:
```suggestion
// ProcessBatches processes `S` concurrently in batches of `batchSize`
```
Note: I often remove types from comments, because the user sees the param names and types in the signature in godoc and all other circumstances. It doesn't help to repeat the types, and it makes it harder to read and internalise quickly.

Reply:
> Because I don't think you can guarantee the goros will execute in the order scheduled.

They are executed sequentially in that individual go routine.

Here's a better example:

```md
List of elements: [E1, E2, E3, ..., E100]

Batch size: 10

Step-by-step process:
1. Divide the list into batches of 10 elements each:
   Batch 1: [E1, E2, E3, ..., E10]
   Batch 2: [E11, E12, E13, ..., E20]
   ...
   Batch 10: [E91, E92, E93, ..., E100]

2. Spawn 10 goroutines to process the batches concurrently:
   Goroutine 1 -> Processes Batch 1
   Goroutine 2 -> Processes Batch 2
   ...
   Goroutine 10 -> Processes Batch 10

3. Within each goroutine:
   - Process elements in the batch sequentially using the `process` function.
   Example for Batch 1:
       Process E1 -> Process E2 -> Process E3 -> ... -> Process E10
```

Reply:
You can't guarantee `process` does it sequentially. You can't even guarantee that `process` does anything from here. Its just that you're doing it sequentially for ws subs for the moment. I agree that comment about sequentially processed should be removed.

Reply:
How about this as an option for the commentary.?
// ProcessBatches splits `S` into batches of `batchSize` and processes each batch concurrently.

There is also an issue when batchSize >= len(S) then there is no concurrency. Should I handle that case?